### PR TITLE
Add initial e-graph implementation

### PR DESCRIPTION
**Pull request by Phantomical**
This PR finally introduces the `EGraph` class. Check out the documentation on the class or [the paper](https://arxiv.org/pdf/2004.03082.pdf) for an explanation of how it works. I haven't really tested this code but it's not used anywhere at this point so it should be fine.

/stack #677

### CODE REVIEW

File: include/caffeine/IR/EGraph.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  bool operator==(const EClass& eclass) const;
+  bool operator!=(const EClass& eclass) const;
+
+  void merge(EClass& eclass);
+
+  friend llvm::hash_code hash_value(const EClass& node);
+};
+
+/**
+ * An implementation of an E-Graph as described in the egg paper.
```
Comment (mishazharov) [inline include/caffeine/IR/EGraph.h:45]: ah yes, the egg paper ;)

File: include/caffeine/Support/Hashing.h [hunk 2]
```diff
@@ -22,7 +22,7 @@ struct LLVMHasher {
 
 #define CAFFEINE_DECL_LLVM_HASHER(type)                                        \
   template <>                                                                  \
-  struct ::std::hash<type> {                                                   \
+  struct std::hash<type> {                                                     \
```
Comment (mishazharov) [inline include/caffeine/Support/Hashing.h:25]: is this new whitespace at the end of the line? might need formatting

**Review by mishazharov — APPROVED** [review]
nits and jokes but otherwise lgtm

File: include/caffeine/IR/EGraph.h (hunk 1 above)
Comment (Phantomical) [reply include/caffeine/IR/EGraph.h:45]: I do link it at the bottom of the comment :P

The paper is actually titled egg though: https://arxiv.org/pdf/2004.03082.pdf

File: include/caffeine/Support/Hashing.h (hunk 2 above)
Comment (Phantomical) [reply include/caffeine/Support/Hashing.h:25]: New whitespace is due to removing the `::` at the start of the name since that's apparently not valid.

**Pull request merged**